[GTK] Adopt libepoxy

zdobersek opened the pull request:
#### 9f70dc3ab09e8c7518c72d667c3782aff91251af
<pre>
[GTK] Adopt libepoxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=248074">https://bugs.webkit.org/show_bug.cgi?id=248074</a>

Reviewed by Carlos Garcia Campos.

Have the GTK port use libepoxy instead of directly including system-provided
graphics libraries headers.

This just brings in libepoxy and uses it. There&apos;s good opportunities of
ripping out different code sections or guarding blocks that have to be
evaluated more thoroughly. The EGL and GL/GLES2 dependencies, already
unused, will also be removed later.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/egl/GLContextEGLWayland.cpp:
* Source/WebCore/platform/graphics/egl/GLContextEGLX11.cpp:
* Source/WebCore/platform/graphics/glx/GLContextGLX.cpp:
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp:
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.cpp:
* Source/WebCore/platform/graphics/x11/XUniquePtr.cpp: Copied from Source/WebCore/platform/graphics/x11/XUniquePtr.h.
(WebCore::XPtrDeleter&lt;__GLXcontextRec&gt;::operator const):
* Source/WebCore/platform/graphics/x11/XUniquePtr.h:
(WebCore::XPtrDeleter&lt;__GLXcontextRec&gt;::operator()): Deleted.
* Source/WebCore/platform/graphics/x11/XUniqueResource.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreWayland.cpp:
* Source/cmake/OptionsGTK.cmake:

Canonical link: <a href="https://commits.webkit.org/256913@main">https://commits.webkit.org/256913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92e42a174e6d614b0b870a9c1e6f46c71b693974

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97182 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6449 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30323 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106700 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166972 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101152 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6726 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35183 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89572 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103389 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102849 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5058 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83809 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32036 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86907 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88098 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/471 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/83542 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/455 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21660 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28450 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4766 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5253 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44157 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/86251 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1695 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40978 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19429 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/x11/XUniquePtr.h:
```diff
@@ -27,13 +27,18 @@
 #if USE(GLX)
+#if USE(LIBEPOXY)
+#include <epoxy/glx.h>
+#else
```
Do we really need this here? It's problematic to include platform gl headers from .h files.

zdobersek replied:
It is needed, but it's avoidable if the template definition is moved into a .cpp file that has to be created.

carlosgcampos commented on Source/WebCore/PlatformGTK.cmake:
```diff
@@ -107,7 +108,7 @@ list(APPEND WebCore_SYSTEM_INCLUDE_DIRECTORIES
     ${UPOWERGLIB_INCLUDE_DIRS}
 )
 
-if (USE_OPENGL)
+if (USE_OPENGL AND NOT USE_LIBEPOXY)
```
Since libepoxy is required, we can just remove the if entirely. And conditional includes in the files that are specific to GTK and/or WPE.

zdobersek replied:
I would like to do this separately, go through `USE_LIBEPOXY` and remove disabled code. But it requires understanding what's specific to GTK and what's also affecting other ports.

zdobersek committed:
[GTK] Adopt libepoxy
https://bugs.webkit.org/show_bug.cgi?id=248074

Reviewed by Carlos Garcia Campos.

Have the GTK port use libepoxy instead of directly including system-provided
graphics libraries headers.

This just brings in libepoxy and uses it. There's good opportunities of
ripping out different code sections or guarding blocks that have to be
evaluated more thoroughly. The EGL and GL/GLES2 dependencies, already
unused, will also be removed later.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/egl/GLContextEGLWayland.cpp:
* Source/WebCore/platform/graphics/egl/GLContextEGLX11.cpp:
* Source/WebCore/platform/graphics/glx/GLContextGLX.cpp:
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp:
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.cpp:
* Source/WebCore/platform/graphics/x11/XUniquePtr.cpp: Copied from Source/WebCore/platform/graphics/x11/XUniquePtr.h.
(WebCore::XPtrDeleter<__GLXcontextRec>::operator const):
* Source/WebCore/platform/graphics/x11/XUniquePtr.h:
(WebCore::XPtrDeleter<__GLXcontextRec>::operator()): Deleted.
* Source/WebCore/platform/graphics/x11/XUniqueResource.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreWayland.cpp:
* Source/cmake/OptionsGTK.cmake:

Canonical link: https://commits.webkit.org/256913@main
